docs/snapshot-versioning.md

```diff
@@ -0,0 +1,96 @@
+|----|----|----|
+| magic_id | 64 | firecracker snapshot, architecture (x86_64/aarch64) and storage version.
+| version  | 64 | the snapshot version number internally 1:1 mapped to specific Firecracker version.
+| state | N*8 | Bincode blob containing the vm state.
+ | crc| 64 | Optional CRC64 sum of magic_id, version and state.
+
+Note: the last 16 bits of `magic_id` encode the storage version which specifies the encoding used for the `version` and `state` fields. The current implementation sets this field to 1 which identifies it as a [Serde bincode](https://github.com/servo/bincode) compatible encoder/decoder.
+
+### Version tolerant ser/de
+Firecracker reads and writes the `state` blob of the snapshot by using per version, separate serialization and deserialization logic. This logic is mostly autogenerated by a Rust procedural macro based on `struct`, `enum` and `union` annotations. Basically, one can say that these structures support versioning. The versioning logic is generated by parsing a structure's history log (encoded using Rust annotations) and emitting rust code.
+
+Versioned serialization and deserialization is divided in 2 translation layers:
+ - structural translator
+ - semantic translator
+
+The structural translator implements the logic to deserialize a structure field by field from any of it's previous versions and recreate the appropriate Rust POD structure at the latest version. It also implements the reverse: given a Rust POD structure it will serialize only the fields present at a specific target version.
+
+The semantic translator works on top of the structural one and its only concern is translating the semantics of serialized/deserialized fields. 
+
+This block diagram illustrates the concept:
+![Versionize](
+images/versionize.png?raw=true
+"Versionize layers")
+
+## Bincode encoding
+During research and prototyping we have considered multiple storage formats. The criteria used for comparing these are: performance, size, rust support, specification, versioning support, community and tooling. Performance, size and rust support are hard requirements while all others will can be the subject of trade offs. 
+
+More info about this comparison can be found [here](https://github.com/firecracker-microvm/firecracker/blob/9d427b33d989c3225d874210f6c2849465941dc0/docs/snapshotting/design.md).
+
+Key benefits of using bincode versus everything else:
+- Minimal snapshot size overhead
+- Minimal CPU overhead 
+- Simple implementation
+
+## Snapshot compatibility
+### Host kernel
+
+The minimum kernel version required by Firecracker snapshots is 4.14. Snapshots can be saved and restored on different kernel versions without issues.
+
+### Device model
+The VM state file links some resources that are external to the snapshot:
+- tap devices by device name
+- block devices by block file path
+
+To succesfully restore a microVM one should check that:
+- tap devices are available and their name match the value in the vmstate file
```

Comment:
```suggestion
- tap devices are available, their name match their original name since that it the value saved in the vmstate file, and they are accessible to the firecracker process where the microVM is being restored
```